src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -625,22 +733,54 @@ evalELam :: Knowledge -> ICtx -> EvalType -> (Symbol, Sort) -> Expr -> EvalST (E
 evalELam γ ctx et (x, s) e = do
     oldPendingUnfoldings <- gets evPendingUnfoldings
     oldEqs <- gets evNewEqualities
+
+    -- We need to declare the variable in the environment 
+    modify $ \st -> st 
+      { evEnv = declareVar x s $ evEnv st }
+
     (e', fe) <- eval (γ { knLams = (x, s) : knLams γ }) ctx et e
     let e2' = simplify γ ctx e'
         elam = ELam (x, s) e
     -- Discard the old equalities which miss the lambda binding
     modify $ \st -> st
       { evPendingUnfoldings = oldPendingUnfoldings
       , evNewEqualities = S.insert (elam, ELam (x, s) e2') oldEqs
+      -- Leaving the scope thus we need to get rid of it
+      , evEnv = deleteVar x $ evEnv st
       }
-    return (elam, fe)
+    return (ELam (x, s) e', fe)
+
+declareVar :: Symbol -> Sort -> SymEnv -> SymEnv
+declareVar x s env = env { seSort = SE $ M.insert x s (seBinds $ seSort env) }
+
+deleteVar :: Symbol -> SymEnv -> SymEnv
+deleteVar x env = env { seSort = SE $ M.delete x (seBinds $ seSort env) }
 
 data RESTParams oc = RP
   { oc   :: OCAlgebra oc Expr IO
   , path :: [(Expr, TermOrigin)]
   , c    :: oc
   }
 
+-- With the support for lambdas in PLE now lambda expression can interact with the 
+-- rewriting feature. But for the moment we don't support them as we can't convert to
+-- REST terms.
+isExprRewritable :: Expr -> Bool
```

Comment:
> I don't know how to make the comment more understandable

I think the comment should make explicit which terms are considered to be rewritable/convertible-to-rest, and which terms are considered not to be. Then we can inspect the implementation to know if it is faithful to the intent.

Reply:
I've looked [here](https://github.com/ucsd-progsys/liquid-fixpoint/blob/4233197fbefa143537d8dc17f6a21be5d88f5d57/src/Language/Fixpoint/Solver/Rewrite.hs#L92)

Reply:
Ok, but there is no need to understand REST. What we would need is a statement like, e.g. "an expression is rewritable if it contains no lambdas."

Then we can check that the implementation correctly flags all expressions containing at least one lambda.

Then someone who understands REST can evaluate one day if this definition of rewritable is overly restrictive.

Reply:
Oh well, I guess it could also say something like: "an expression is rewritable if it is in the domain of Language.Fixpoint.Solver.Rewrite.convert". Then the documentation of `convert` would need to refer back to `isExprRewritable` to keep both functions in sync.